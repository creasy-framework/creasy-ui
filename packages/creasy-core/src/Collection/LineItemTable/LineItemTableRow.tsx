import React, {FunctionComponent, Fragment, ReactNode} from 'react';
import Table from '../Table/Table';

export interface LineItemTableColumnConfig {
  columnName: string;
  textWrap?: 'nowrap' | 'break-all' | 'break-word';
  className?: string;
  hideLabel?: boolean;
}

interface Props {
  columnsConfig: LineItemTableColumnConfig[];
}

export interface LineItemTableRowComponent extends FunctionComponent<Props> {}

const LineItemTableRow: LineItemTableRowComponent = ({ children, columnsConfig }) => {
  const columns = children instanceof Array ? children : [children];
  return (
    <Fragment>
      {
        columns.map((columnElement: ReactNode, index) => (
          <Table.RowItem
            key={columnsConfig[index].columnName}
            {...columnsConfig[index]}
          >
            {columnElement}
          </Table.RowItem>
        ))
      }
    </Fragment>
  );
};

export default LineItemTableRow;
