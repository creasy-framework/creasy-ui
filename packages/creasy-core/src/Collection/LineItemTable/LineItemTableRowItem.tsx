import React, { FunctionComponent, Fragment } from 'react';

interface Props {}

export interface LineItemTableRowItemComponent extends FunctionComponent<Props> {}

const LineItemTableRowItem: LineItemTableRowItemComponent = ({ children }) => {
  return (
    <Fragment>
      {children}
    </Fragment>
  );
};

export default LineItemTableRowItem;
