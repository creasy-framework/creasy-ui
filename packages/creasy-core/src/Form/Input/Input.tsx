import * as React from 'react';
import { ReactElement, SyntheticEvent } from 'react';
import Cleave from 'cleave.js/react';
import classNames from 'classnames';
import FieldGroup from '../FieldGroup/FieldGroup';

interface Props {
  value?: string | number;
  label?: ReactElement | string;
  disabled?: boolean;
  readOnly?: boolean;
  isRequired?: boolean;
  labelAccessory?: ReactElement;
  labelPlacement?: 'left' | 'top';
  onChange?: (e: SyntheticEvent) => void;
  onBlur?: (e: SyntheticEvent) => void;
  onKeyDown?: (e: SyntheticEvent) => void;
  onFocus?: (e: SyntheticEvent) => void;
  size?: 'xs' | 's' | 'm' | 'l' | 'xl';
  placeholder?: string;
  options?: object;
  type?: 'outlined' | 'underlined';
  left?: ReactElement;
  right?: ReactElement;
  className?: string;
  renderInput?: (inputProps: any) => ReactElement;
}

class Input extends React.Component<Props> {
  state = {
    hasFocus: false,
  };

  static defaultProps = {
    type: 'outlined',
    labelPlacement: 'left'
  }

  private onInputFocus = (e: SyntheticEvent) => {
    const { onFocus } = this.props;
    this.setState({
      hasFocus: true,
    });
    onFocus && onFocus(e);
  }

  private onInputBlur = (e: SyntheticEvent) => {
    const { onBlur } = this.props;
    this.setState({
      hasFocus: false,
    });
    onBlur && onBlur(e);
  }

  render() {
    const {
      type = 'outlined',
      label,
      isRequired,
      labelAccessory,
      labelPlacement,
      options,
      size,
      left,
      right,
      className,
      value = '',
      renderInput,
      ...otherProps
    } = this.props;

    const { hasFocus } = this.state;
    const inputProps = {
      value,
      onChange: () => {},
      ...otherProps,
      className: 'creasy-input__control',
      onFocus: this.onInputFocus,
      onBlur: this.onInputBlur,
    };

    const inputView = options ? (
      <Cleave
        {...inputProps}
        options={options}
      >
      </Cleave>
    ) : <input {...inputProps} />;

    return (
      <FieldGroup
        label={label}
        isRequired={isRequired}
        labelAccessory={labelAccessory}
        size={size}
        labelPlacement={labelPlacement}
      >
        <div
          className={
            classNames(
              'creasy-input',
              hasFocus && 'creasy-input--active',
              type && `creasy-input--type-${type}`,
              className,
            )
          }>
          {
            left && <span className="creasy-input__left">{left}</span>
          }
          {
            renderInput ? renderInput(inputProps) : inputView
          }
          {
            right && <span className="creasy-input__right">{right}</span>
          }
        </div>
      </FieldGroup>
    );
  }
}

export default Input;
