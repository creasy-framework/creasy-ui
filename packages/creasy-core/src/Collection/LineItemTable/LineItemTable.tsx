import React, { FunctionComponent, ReactNode } from 'react';
import shortid from 'shortid';

import Table from '../Table/Table';
import Icons from '../../Graphic/Icons';

import LineItemTableRow, { LineItemTableRowComponent, LineItemTableColumnConfig } from './LineItemTableRow';
import LineItemTableRowItem, { LineItemTableRowItemComponent } from './LineItemTableRowItem';
import IconButton from '../../Form/IconButton/IconButton';
import classNames from "classnames";

interface Props {
  data: any[];
  columnsConfig: LineItemTableColumnConfig[];
  onUpdateRow: (index: number, ...args: any) => void;
  onAddRow?: (...args: any) => void;
  onRemoveRow?: (index: number) => void;
  children: (index: number, rowData: any, onUpdateRow: (...args: any) => void) => ReactNode;
}

interface LineItemTableComponent extends FunctionComponent<Props> {
  Row: LineItemTableRowComponent;
  Item: LineItemTableRowItemComponent;
}

const LineItemTable: LineItemTableComponent = ({
 data,
 columnsConfig,
 children,
 onUpdateRow,
 onAddRow,
 onRemoveRow,
}) => {
  data.forEach((row: any) => {
    if (!row.id) {
      row.id = shortid.generate();
    }
  });

  const newRow = { id: shortid.generate() };

  const rows = onAddRow ? [...data, newRow] : [...data];

  const onChange = (index: number, rowData: any, ...args: any) => {
    if (index === data.length) {
      onAddRow && onAddRow(rowData, ...args);
    } else {
      onUpdateRow && onUpdateRow(index, ...args);
    }
  };

  return (
    <Table className={classNames('creasy-line-item-table', !!onAddRow && 'creasy-line-item-table--appendable')}>
      <Table.Header>
        {
          columnsConfig.map(({ columnName, ...otherProps }) => (
            <Table.HeaderItem
              key={columnName}
              {...otherProps}
            >
              {columnName}
            </Table.HeaderItem>
          ))
        }
        {
          onRemoveRow && <Table.RowItem className="creasy-line-item-table__remove"/>
        }
      </Table.Header>
      <Table.Body>
        {
          rows.map((rowData, index) => (
            <Table.Row key={rowData.id} className="creasy-line-item-table__row">
              {children(index, rowData, (...args) => onChange(index, rowData, ...args))}
              {
                onRemoveRow && (
                  <Table.RowItem textWrap="break-all" className="creasy-line-item-table__remove" columnName="Remove" hideLabel>
                    <IconButton
                      className="creasy-line-item-table__remove-button"
                      onClick={() => onRemoveRow(index)}
                    >
                      <Icons.Trash />
                    </IconButton>
                  </Table.RowItem>
                )
              }
            </Table.Row>
          ))
        }
      </Table.Body>
    </Table>
  );
};

LineItemTable.Row = LineItemTableRow;
LineItemTable.Item = LineItemTableRowItem;

export default LineItemTable;
