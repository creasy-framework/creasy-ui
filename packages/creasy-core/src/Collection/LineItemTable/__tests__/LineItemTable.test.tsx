import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import LineItemTable from '../LineItemTable';
import Input from '../../../Form/Input/Input';

jest.mock('shortid', () => ({
  generate: () => '123',
}));

describe('LineItemTable', () => {
  let wrapper: ShallowWrapper;

  const data = [
    { id: '1', name: 'CHENG LI', team: 'Customer Service', date: '02/03/2020' },
    { id: '2', name: 'SUN WEI', team: 'Customer Service', date: '02/03/2020' },
    { id: '3', name: 'LIPING ZHAO', team: 'Customer Service', date: '02/03/2020' },
  ];

  const columnsConfig = [
    { columnName: 'Name' },
    { columnName: 'Team' },
    { columnName: 'On boarding date', style: { width: 200 } },
  ];

  beforeAll(() => {
    wrapper = shallow(
      <LineItemTable
        columnsConfig={columnsConfig}
        data={data}
      >
        {(index, data) => (
          <LineItemTable.Row columnsConfig={columnsConfig}>
            <LineItemTable.Item>
              <Input value={data.name}/>
            </LineItemTable.Item>
            <LineItemTable.Item>
              <Input value={data.team}/>
            </LineItemTable.Item>
            <LineItemTable.Item>
              <Input value={data.date}/>
            </LineItemTable.Item>
          </LineItemTable.Row>
        )}
      </LineItemTable>,
    );
  });

  it('should render properly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
