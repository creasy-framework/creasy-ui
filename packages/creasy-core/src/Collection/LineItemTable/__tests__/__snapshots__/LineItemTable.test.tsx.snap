// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineItemTable should render properly 1`] = `
<Table
  className="creasy-line-item-table"
>
  <Header>
    <HeaderItem
      key="Name"
    >
      Name
    </HeaderItem>
    <HeaderItem
      key="Team"
    >
      Team
    </HeaderItem>
    <HeaderItem
      key="On boarding date"
      style={
        Object {
          "width": 200,
        }
      }
    >
      On boarding date
    </HeaderItem>
  </Header>
  <Body>
    <Row
      className="creasy-line-item-table__row"
      key="1"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="CHENG LI"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
    </Row>
    <Row
      className="creasy-line-item-table__row"
      key="2"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="SUN WEI"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
    </Row>
    <Row
      className="creasy-line-item-table__row"
      key="3"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="LIPING ZHAO"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            labelPlacement="left"
            type="outlined"
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
    </Row>
  </Body>
</Table>
`;
